ecl/hql/hqlgram2.cpp

```diff
@@ -9271,6 +9271,85 @@ bool HqlGram::checkAllowed(const attribute & errpos, const char *category, const
 }
 
+bool HqlGram::checkDFSfields(IHqlExpression *dfsRecord, IHqlExpression *defaultRecord, const attribute& errpos)
+{
+    if (dfsRecord==defaultRecord)
+        return true;
+    bool ret = true;
+    IHqlSimpleScope *srcScope = dfsRecord->querySimpleScope();
+    ForEachChild(idx, defaultRecord)
+    {
+        IHqlExpression *subfield = defaultRecord->queryChild(idx);
+        switch (subfield->getOperator())
+        {
+        case no_record:
+        case no_ifblock:
+        case no_field:
+            {
+                OwnedHqlExpr match = srcScope->lookupSymbol(subfield->queryId());
+                if (!match)
+                {
+                    reportError(HQLERR_DFSlookupIncompatible, errpos, "Default record does not match resolved record - field %s not found", str(subfield->queryId()));
+                    ret = false;
+                }
+                else if (subfield->getOperator() != match->getOperator())
+                {
+                    reportError(HQLERR_DFSlookupIncompatible, errpos, "Default record does not match resolved record - field %s incompatible", str(subfield->queryId()));
+                    ret = false;
+                }
+                else if (subfield->queryRecord() && match->queryRecord())
+                {
+                        ret = checkDFSfields(match->queryRecord(), match->queryRecord(), errpos) && ret;
+                }
+                else if (match->queryRecord() || subfield->queryRecord())
+                {
+                    reportError(HQLERR_DFSlookupIncompatible, errpos, "Default record does not match resolved record - field %s incompatible", str(subfield->queryId()));
+                    ret = false;
+                }
+                else
+                {
+                    if (!checkCompatible(match->queryType(), subfield->queryType(), errpos, false))
+                    {
+                        reportError(HQLERR_DFSlookupIncompatible, errpos, "Default record does not match resolved record - field %s incompatible", str(subfield->queryId()));
+                        ret = false;
+                    }
+                }
+            }
+        }
+    }
+    return ret;
+}
+
+IHqlExpression * HqlGram::lookupDFSlayout(const attribute &err, IHqlExpression *_name, IHqlExpression *_default, IHqlExpression *_lookupAttr, bool isOpt)
+{
+    OwnedHqlExpr defaultRecord(_default);
+    OwnedHqlExpr lookupAttr(_lookupAttr);
+    OwnedHqlExpr nameExpr(_name);
+    if (lookupAttr)
+    {
+        IHqlExpression * value = lookupAttr->queryChild(0);
+        if (!getBoolValue(value, true))
+            return defaultRecord.getClear();  // LOOKUP(FALSE) was specified
+    }
+    if (lookupCtx.queryParseContext().codegenCtx)
+    {
+        if (!nameExpr->isConstant())
+            reportError(ERR_EXPECTED_CONST, err, "LOOKUP attribute requires a constant filename");
+        StringBuffer lookupName;
+        nameExpr->queryValue()->getStringValue(lookupName);
```

Comment:
I think I was thinking that reportError did not return, but that's probably not true...